File: C/GildedRose.c
#include <stdbool.h>
#include <string.h>
#include "GildedRose.h"
#include <stdio.h>

#define MAX_QUALITY 50
#define MIN_QUALITY  0

Item*
init_item(Item* item, const char *name, int sellIn, int quality)
{
    item->sellIn = sellIn;
    item->quality = quality;
    item->name = strdup(name);
    
    return item;
}

extern char* 
print_item(char* buffer, Item* item) 
{
    sprintf(buffer, "%s, %d, %d", item->name, item->sellIn, item->quality);
}

bool is_brie(Item item)
{
    return (0 == strcmp(item.name, "Aged Brie"));
}

bool is_passes(Item item)
{
    return (0 == strcmp(item.name, "Backstage passes to a TAFKAL80ETC concert"));
}

bool is_sulfuras(Item item)
{
    return (0 == strcmp(item.name, "Sulfuras, Hand of Ragnaros"));
}

void increment_quality(Item *item)
{
    if (item->quality < MAX_QUALITY)
        ++item->quality;
}

void
update_quality(Item items[], int size) 
{
    int i;
    
    for (i = 0; i < size; i++)
    {
        if (!is_brie(items[i]) && !is_passes(items[i]))
        {
            if (items[i].quality > MIN_QUALITY && !is_sulfuras(items[i]))
            {
                --items[i].quality;
            }
        }
        else
        {
            increment_quality(&items[i]);

            if (is_passes(items[i]))
            {
                if (items[i].sellIn < 11)
                    increment_quality(&items[i]);

                if (items[i].sellIn < 6)
                    increment_quality(&items[i]);
            }
        }

        if (!is_sulfuras(items[i]))
        {
            --items[i].sellIn;
        }

        if (items[i].sellIn < 0)
        {
            if (!is_brie(items[i]))
            {
                if (!is_passes(items[i]))
                {
                    if (items[i].quality > MIN_QUALITY && !is_sulfuras(items[i]))
                    {
                        --items[i].quality;
                    }
                }
                else
                {
                    items[i].quality = 0;
                }
            }
            else
            {
                increment_quality(&items[i]);
            }
        }
    }
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            MalleableItem.factory(item).update()


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)


class MalleableItem:
    def __init__(self, item):
        self.item = item

    @staticmethod
    def factory(item):
        if item.name == "Sulfuras, Hand of Ragnaros":
            return Sulfuras(item)
        elif item.name == "Aged Brie":
            return AgedBrie(item)
        elif item.name == "Backstage passes to a TAFKAL80ETC concert":
            return BackstagePasses(item)
        elif item.name == "Conjured Mana Cake":
            return Conjured(item)
        else:
            return MalleableItem(item)

    def decrease_sell_in(self):
        item = self.item

        item.sell_in = item.sell_in - 1

    def decrease_quality(self):
        item = self.item

        if item.quality > 0:
            item.quality = item.quality - 1

    def increase_quality(self):
        item = self.item

        if item.quality < 50:
            item.quality = item.quality + 1

    def update(self):
        item = self.item

        self.decrease_quality()

        self.decrease_sell_in()

        if item.sell_in < 0:
            self.decrease_quality()


class Sulfuras(MalleableItem):
    def decrease_quality(self):
        pass

    def decrease_sell_in(self):
        pass


class AgedBrie(MalleableItem):
    def update(self):
        self.increase_quality()

        self.decrease_sell_in()

        if self.item.sell_in < 0:
            self.increase_quality()


class BackstagePasses(MalleableItem):
    def update(self):
        self.increase_quality()
        if self.item.sell_in < 11:
            self.increase_quality()
        if self.item.sell_in < 6:
            self.increase_quality()

        self.decrease_sell_in()

        if self.item.sell_in < 0:
            self.item.quality = 0


class Conjured(MalleableItem, object):
    def decrease_quality(self):
        super(Conjured, self).decrease_quality()
        super(Conjured, self).decrease_quality()
